server/aesdsocket.c

```diff
@@ -200,105 +363,155 @@ int main(int argc, char *argv[])
         if (new_fd == -1) {
             printf("in accept fail with int\n");
             if(errno == EINTR)
             {
-                break;
+                goto join_threads;
             }
             perror("accept");
             continue;
         }
-        reading = true;
-        inet_ntop(their_addr.ss_family,
-            get_in_addr((struct sockaddr *)&their_addr),
-            s, sizeof s);
-        printf("server: got connection from %s\n", s);
-
-        int local_fd = open("/var/tmp/aesdsocketdata", O_CREAT | O_RDWR | O_APPEND, S_IRWXU |  S_IRWXG | S_IRWXO );
-        printf("file opened\n");
-        int max_buf_size = BUF_SIZE;
-        char *buf = (char *)malloc(max_buf_size * sizeof(char));
-        int return_idx = 0;
-        int num_of_relloc = 0;
-        bool ret_found = false;
-
-        while ( !ret_found ) //return not found
+
+
+
+        printf("before malloc\n");
+        thread_data_t *thread_list = NULL;
+        thread_list = malloc(sizeof(thread_data_t));
+        thread_list->thread_complete_success = false;
+        thread_list->sfd = new_fd;
+        thread_list->local_fd = local_fd;
+        
+
+        int ret = pthread_create(&thread_list->tid, NULL, thread_foo, thread_list);
+        if (!ret)
         {
-            if(recv(new_fd, buf + (num_of_relloc * BUF_SIZE), BUF_SIZE, 0) == - 1){
-                //log errors
-                printf("recv() failed\n");
-            }
+            SLIST_INSERT_HEAD(&head, thread_list, entries);
+            printf("Thread created\n");
+            num_thread++;
+        }
+        else{
+            printf("Thread not created\n");
+            perror("pthread_create()");
 
-            return_idx = search_for_return(buf + (num_of_relloc * BUF_SIZE), BUF_SIZE);
-            if(return_idx == -1)
+
+            //need to clean up
+        }
+        join_threads:
+        printf("Joining threads");
+        thread_data_t *e = NULL;
+        thread_data_t* next = NULL;
+        SLIST_FOREACH_SAFE(e, &head, entries,next)
+        {
+            if(e->thread_complete_success == true)
             {
-                max_buf_size += BUF_SIZE;
-                buf = (char *)realloc(buf, max_buf_size * sizeof(char));
-                num_of_relloc++;
+                pthread_join(e->tid, NULL);
+                printf("removed\n");
+                //need to remove node 
+                SLIST_REMOVE(&head, e, thread_data_s, entries);
+                num_thread--;
+                free(e);
             }
-            else
+            e = NULL;
+        }
+        printf("Num thread %d\n",num_thread);
+        if(num_thread == 0)
+        {
+            if(time_elasped > time_written)
             {
-                printf("writing to file\n");
-                write_to_file(local_fd, buf, max_buf_size - (BUF_SIZE - return_idx));
-                ret_found = true;
-                break;
+                write_to_file(local_fd, time_buff, strlen(time_buff));
```

Comment:
@PeterBraganza how do you ensure `time_buff` isn't overwritten by the signal handler during this call?